### ISSUE TITLE
Computation of posteriors and null pointer exception 

### ISSUE THREAD

**Issue by bussfromspace** [busarobi/XMLC#26]
Hi,
I do not know if this is the error related to each other, but when I try to compute the posteriors for test and validation set, the code does not terminate in the terminal. It seems that it executes the writePosteriorsToFile() function and I can see the counter inside the function is increasing, but I do not see that if the function fully returns from the terminal or not. 
 
Also, when I try to predict labels by using the trained model and the macro F-measure optimized thresholds, it gives null pointer exception error like this:

t31300-la024:XMLC atlib1$ java -Xmx12G -jar XMLC_PLT-jar-with-dependencies.jar -test ./examples/rcv1_test.config
(2018-05-12 10:48:44,269) INFO  [main] [Run.main:21] - --> Use case: -test
(2018-05-12 10:48:44,273) INFO  [main] [Run.main:22] - --> Config file./examples/rcv1_test.config
(2018-05-12 10:48:44,278) INFO  [main] [ReadProperty.readProperty:14] - Reading property file...
(2018-05-12 10:48:44,283) INFO  [main] [ReadProperty.readProperty:27] - Done.
(2018-05-12 10:48:44,707) INFO  [main] [LearnerManager.loadmodel:194] - Working Directory = /Users/atlib1/XMLC
(2018-05-12 10:48:44,717) INFO  [main] [AbstractLearner.learnerFactory:72] - --> Learner: PLT
#####################################################
#### Learner: PLT
(2018-05-12 10:48:44,735) INFO  [main] [PLT.<init>:82] - #### gamma: 1.0
(2018-05-12 10:48:44,735) INFO  [main] [PLT.<init>:86] - #### lambda: 1.0
(2018-05-12 10:48:44,735) INFO  [main] [PLT.<init>:90] - #### epochs: 30
(2018-05-12 10:48:44,736) INFO  [main] [PLT.<init>:94] - #### Hasher: Mask
(2018-05-12 10:48:44,736) INFO  [main] [PLT.<init>:98] - #### Number of ML hashed features: 50000000
(2018-05-12 10:48:44,737) INFO  [main] [PLT.<init>:103] - #### k (order of the tree): 2
(2018-05-12 10:48:44,737) INFO  [main] [PLT.<init>:107] - #### tree type Complete
(2018-05-12 10:48:44,738) INFO  [main] [PLT.<init>:111] - #### tree file name null
#####################################################
(2018-05-12 10:48:44,738) INFO  [main] [LearnerManager.loadmodel:209] - #################################################### 
(2018-05-12 10:48:44,739) INFO  [main] [LearnerManager.loadmodel:210] - Loading model file from ./examples/model_ontrain.model
(2018-05-12 10:49:06,285) INFO  [main] [MaskHasher.<init>:23] - #####################################################
(2018-05-12 10:49:06,289) INFO  [main] [MaskHasher.<init>:24] - #### Mask hash
(2018-05-12 10:49:06,289) INFO  [main] [MaskHasher.<init>:25] - #### Num. of hashed features: 1318286524416
(2018-05-12 10:49:06,290) INFO  [main] [MaskHasher.<init>:26] - #####################################################
(2018-05-12 10:49:06,292) INFO  [main] [LearnerManager.loadmodel:212] - Parameters of the loaded model 
(2018-05-12 10:49:06,293) INFO  [main] [AbstractLearner.printParameters:64] - Number of labels: 2456
(2018-05-12 10:49:06,293) INFO  [main] [AbstractLearner.printParameters:65] - Number of features: 47236
(2018-05-12 10:49:06,294) INFO  [main] [PLT.printParameters:119] - #### gamma: 0.1
(2018-05-12 10:49:06,295) INFO  [main] [PLT.printParameters:120] - #### lambda: 1.0E-6
(2018-05-12 10:49:06,295) INFO  [main] [PLT.printParameters:121] - #### epochs: 10
(2018-05-12 10:49:06,295) INFO  [main] [PLT.printParameters:122] - #### Hasher: Mask
(2018-05-12 10:49:06,296) INFO  [main] [PLT.printParameters:123] - #### Number of ML hashed features: 268435456
(2018-05-12 10:49:06,296) INFO  [main] [PLT.printParameters:124] - #### k (order of the tree): 32
(2018-05-12 10:49:06,296) INFO  [main] [PLT.printParameters:125] - #### tree type: HuffmanTree
(2018-05-12 10:49:06,296) INFO  [main] [PLT.printParameters:126] - #### tree file: ./examples/huffman_tree.txt
(2018-05-12 10:49:06,297) INFO  [main] [LearnerManager.loadmodel:214] - #################################################### 
Exception in thread "main" java.lang.NullPointerException
	at Learner.PLT.setThresholds(PLT.java:500)
	at run.LearnerManager.forecast(LearnerManager.java:129)
	at run.Run.main(Run.java:45)

**Comment by mmcelikok** [comment]
Reproduced. writePosteriorsToFile() returns to outputPosteriors(), however outputPosteriors() doesn't return to main. Thread dump reveals it's waiting for the ArrayBlockingQueue of OnlineDataManager:

"Thread-5" #15 prio=5 os_prio=0 tid=0x00007fec1c1e9000 nid=0x18d6 waiting on condition [0x00007fec02bc1000]
   java.lang.Thread.State: WAITING (parking)
    at sun.misc.Unsafe.park(Native Method)
    - parking to wait for  <0x00000006e4a51068> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
    at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
    at java.util.concurrent.ArrayBlockingQueue.put(ArrayBlockingQueue.java:353)
    at IO.OnlineDataManager$ReaderThread.run(OnlineDataManager.java:142)
    at java.lang.Thread.run(Thread.java:748)
